### PR TITLE
Smoke position & animation speed

### PR DESCRIPTION
**Pull request by spacek531**
extending the abilities of objects slightly to move the steam effect of steam trains, and (hopefully) tune the speed that animations play at.

The attached object has properties set to utilize the features
[TESTDAT.zip](https://github.com/OpenRCT2/OpenRCT2/files/4455317/TESTDAT.zip)

### CODE REVIEW

File: src/openrct2/object/ObjectJsonHelpers.h [hunk 1]
```diff
@@ -29,6 +29,7 @@ namespace ObjectJsonHelpers
     float GetFloat(const json_t * obj, const std::string &name, const float &defaultValue = 0);
     std::vector<std::string> GetJsonStringArray(const json_t * arr);
     std::vector<int32_t> GetJsonIntegerArray(const json_t * arr);
+    std::vector<float_t> GetJsonRealArray(const json_t * arr);
```
Comment (janisozaur) [inline src/openrct2/object/ObjectJsonHelpers.h:32]: Please don't use floating point values for code that touches logic, different compilers and different optimisation flags will lead to divergent results.

File: src/openrct2/ride/Vehicle.h [hunk 2]
```diff
@@ -282,6 +284,9 @@ enum {
     VEHICLE_ENTRY_ANIMATION_MULTI_DIM_COASTER
 };
 
+#define ANIMATION_SPEED_MODIFIER_COEFFICIENT 32
```
Comment (Gymnasiast) [inline src/openrct2/ride/Vehicle.h:287]: All new code should use the C++ counterpart: constexpr.

```suggestion
constexpr const uint8 ANIMATION_SPEED_MODIFIER_COEFFICIENT = 32;
```

File: src/openrct2/ride/Vehicle.h [hunk 3]
```diff
@@ -282,6 +284,9 @@ enum {
...
+#define STEAM_EFFECT_MODIFIER_COEFFICIENT 32
```
Comment (Gymnasiast) [inline src/openrct2/ride/Vehicle.h:288]:
```suggestion
constexpr const uint8 STEAM_EFFECT_MODIFIER_COEFFICIENT = 32;
```
```

File: src/openrct2/object/RideObject.cpp [hunk 4]
```diff
@@ -434,7 +434,11 @@ void RideObject::ReadLegacyVehicle(
     vehicle->effect_visual = stream->ReadValue<uint8_t>();
     vehicle->draw_order = stream->ReadValue<uint8_t>();
     vehicle->num_vertical_frames_override = stream->ReadValue<uint8_t>();
-    stream->Seek(4, STREAM_SEEK_CURRENT);
+    stream->Seek(1, STREAM_SEEK_CURRENT);
+    vehicle->animation_speed_modifier = stream->ReadValue<uint8_t>();
+    vehicle->steam_effect_modifier[0] = stream->ReadValue<uint8_t>();
+    vehicle->steam_effect_modifier[1] = stream->ReadValue<uint8_t>();
+//    stream->Seek(0, STREAM_SEEK_CURRENT);
```
Comment (IntelOrca) [inline src/openrct2/object/RideObject.cpp:441]: New fields should only be added to the JSON format so this needs to be reverted.

File: src/openrct2/ride/Vehicle.h [hunk 5]
```diff
@@ -71,11 +71,13 @@ struct rct_ride_entry_vehicle {
...
-    uint8_t draw_order;
+    uint8_t effect_visual;            // 0x5E , 0x78
+    uint8_t draw_order;               // 0x5F , 0x79
     uint8_t num_vertical_frames_override; // 0x60 , 0x7A, A custom number that can be used rather than letting RCT2 determine it. Needs the VEHICLE_ENTRY_FLAG_OVERRIDE_NUM_VERTICAL_FRAMES flag to be set.
-    uint8_t peep_loading_waypoint_segments; // 0x61 new
-    uint8_t pad_62[6] = {};          // 0x62 , 0x7B
+    uint8_t peep_loading_waypoint_segments; // 0x61, 0x7B new
+    int8_t animation_speed_modifier;  // 0x62 , 0x7C, A multiplier to change animation speed on most animation types
+    int8_t steam_effect_modifier[2] = {}; // 0x63, 0x7D, A multiplier to change smoke particle effect positions in lateral and vertical directions
+    uint8_t pad_65[3] = {};           // 0x65 , 0x7E
```
Comment (IntelOrca) [inline src/openrct2/ride/Vehicle.h:80]: You can remove this pad field.

File: src/openrct2/object/ObjectJsonHelpers.h (hunk 1 above)
Comment (spacek531) [reply src/openrct2/object/ObjectJsonHelpers.h:32]: this should be resolved now

**Comment by spacek531** [comment]
rebasing in progress

**Pull request closed without merging**